README.md
# RailsCountrySelect
[![Gem Version](https://badge.fury.io/rb/rails_country_select.png)](http://badge.fury.io/rb/rails_country_select) [![Code Climate](https://codeclimate.com/github/rderoldan1/rails_country_select.png)](https://codeclimate.com/github/rderoldan1/rails_country_select) [![Build Status](https://travis-ci.org/rderoldan1/rails_country_select.png?branch=master)](https://travis-ci.org/rderoldan1/rails_country_select) [![Coverage Status](https://coveralls.io/repos/rderoldan1/rails_country_select/badge.png?branch=master)](https://coveralls.io/r/rderoldan1/rails_country_select?branch=master)

Rails 3 gem for Select tag in order to choose a country in your forms, provides 4 options.

*This project is based on Rails 2 plugin created by LukeCarrier [rails-country-select](https://github.com/rderoldan1/rails-country-select/blob/master/lib/country_definitions.rb)*

1. Country Name.
2. Cod
3. ISO cod 2 chars ('CO', 'AF', 'AX', 'AL')
4. ISO cod 3 chars ('COL', 'AFG', 'ALA', ALB')

Also includes a Helper method in order to get all country info.

## Installation

Add this line to your application's Gemfile:

    gem 'rails_country_select'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rails_country_select

## Usage

### Basic
* value : country cod number
* text  : country name

```rails
<%= form_for @user do |f| %>
    <%= f.country_select :country %>
<% end %>
```

HTML generated
```html
 <select id="user_country" name="user[country]">
    <option value="4">Afghanistan</option>
    <option value="248">Albania</option>
    <option value="8">Ãland Islands</option>
    <option value="12">Algeria</option>
```

### Advanced
* values, keys : [:names, :nums, :alpha2s, :alpha3s]
* selected      : set default value
* prompt       : message

```rails
<%= form_for @user do |f| %>
    <%= f.country_select :country, {:keys => :names, :values => :alpha2s, :selected => "CO"}, :prompt => "Choose a country"
<% end %>
```

HTML generated

```html
<select id="user_country" name="user[country]" prompt="Choose a country" value="CO">
    <option value="AF">Afghanistan</option>
    <option value="AL">Albania</option>
    <option value="AX">Ãland Islands</option>
    <option value="DZ">Algeria</option>
```


### Show country info in array
A helper method is available to get all the country info in the views, it return an array, the usage is.

Find by name
```ruby
country(:name => "Colombia")
[170, "Colombia", "CO", "COL"]
```

Find by Cod
```ruby
country(:num => 12)
[12, "Algeria", "DZ", "DZA"]
```

Find by Alpha2s
```ruby
country(:alpha2s => "AR")
[32, "Argentina", "AR", "ARG"]
```

Find by Alpha3s
```ruby
country(:alpha3s => "USA")
[840, "United States", "US", "USA"]
```

Also you can mix all the options hash, example
```ruby
country(:alpha3s => "USA", :alpha2s => "US", :num => 840, :name => "United States")
[840, "United States", "US", "USA"]
```

#### Show specific country info
All the methods receive a hash options :name, :cod, :alpha2s, :alpha3s
```ruby
country_name :num => "170"
=> "Colombia"

country_cod :name => "Colombia"
=> 170
country_alpha2s :num => "170"
=> "CO"

country_alpha3s :num => "170"
=> "COL"
```

#### Usage in views

Assuming that your User model store country cod.
```rails
<p>Country Cod: <%= country(:num => @user.country_num)[0] %><p>
<p>Country Name: <%= country(:num => @user.country_num)[1] %><p>
<p>Country Apha2s: <%= country(:num => @user.country_num)[2] %><p>
<p>Country Apha3s: <%= country(:num => @user.country_num)[3] %><p>

<p>Country Cod: <%= country_cod(:num => @user.country_num) %><p>
<p>Country Name: <%= country_name(:num => @user.country_num) %><p>
<p>Country Apha2s: <%= country_alpha2s(:num => @user.country_num) %><p>
<p>Country Apha3s: <%= country_alpha3s(:num => @user.country_num) %><p>

```

### Locales

By default all county names are in english, but you can change this behavior by include country names in
your locale file under `countries:` scope

```yml
es:
 countries:
    AC: "Isla de la Ascensión"
    AD: "Andorra"
    AE: "Emiratos Árabes Unidos"
    AF: "Afganistán"
    AG: "Antigua y Barbuda"
    AI: "Anguila"
    AL: "Albania"
    AM: "Armenia"
```

Please look at [@onomojo](https://github.com/onomojo) [repo](https://github.com/onomojo/i18n-country-translations/tree/master/rails/locale) for all translations




## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG
V1.0.0
========
* First stable version
* Find specific country info helpers


V0.2.2
========
* Country Name Locales

V.0.1.0
========
* Helper method in order to get country info.

V0.0.1
========
* First Version, form tag country_select.



spec/helper_spec.rb
require 'rspec'
require 'spec_helper'

describe 'Helper methods' do

  it 'should find country name' do
    a = country_name(:num => "170")
    a.class.should be String
    a.should eq "Colombia"
    a.should_not be_nil
  end

  it 'should find country alpha2s' do
    a = country_alpha2s(:num => "170")
    a.class.should be String
    a.should eq "CO"
    a.should_not be_nil
  end

  it 'should find country alpha3s' do
    a = country_alpha3s(:num => "170")
    a.class.should be String
    a.should eq "COL"
    a.should_not be_nil
  end

  it 'should find country number' do
    a = country_cod(:name => "Colombia")
    a.class.should be Fixnum
    a.should eq 170
    a.should_not be_nil
  end

  it 'should find country info' do
    a = country(:name => "Colombia")
    a.class.should be Array
    a.count.should eq 4
    a[0].should eq 170
    a[0].class.should eq Fixnum
    a.should_not be_nil
  end
end

spec/spec_helper.rb
require 'rubygems'
require 'bundler/setup'
require 'coveralls'

Coveralls.wear! 'rails'

require File.expand_path('../../lib/rails_country_select', __FILE__)
require File.expand_path('../../lib/country_definitions', __FILE__)

require 'action_view'
require 'action_view/helpers'
require 'i18n'

include ActionView::Helpers
include ActionView::Helpers::FormOptionsHelper
include ActionView::Helpers::FormHelper
include ActionView::Helpers::TagHelper
RSpec.configure do |config|
end


spec/version_spec.rb
require 'rspec'
require 'spec_helper'

describe 'Version' do
  it 'should be the version number' do
    a = RailsCountrySelect::VERSION
    a.class.should be(String)
    a.should_not be_nil
  end
end


lib/rails_country_select.rb
# encoding: utf-8
require "rails_country_select/version"
require File.expand_path('../country_definitions', __FILE__)

#if defined?(Rails) && defined?(ActionView)
  module ActionView
    module Helpers
      autoload :CountryHelper, 'rails_country_select/helpers/rails_country_select'
      module FormOptionsHelper
        def country_select(object, method, options = {}, html_options = {})
          options[:keys]    = :names unless options.has_key?(:keys)
          options[:values]  = :nums  unless options.has_key?(:values)

          potential = {
              :names   => translate_countries,
              :nums    => COUNTRY_NUMS,
              :alpha2s => COUNTRY_ALPHA2S,
              :alpha3s => COUNTRY_ALPHA3S
          }
          
          select_options = potential[options[:keys]].zip(potential[options[:values]])
          InstanceTag.new(object, method, self, options.delete(:object)).to_select_tag(select_options, options, html_options)
        end
      end

      class FormBuilder
        def country_select(method, options = {}, html_options = {})
          @template.country_select(@object_name, method, options.merge({:object => @object}), html_options)
        end
      end

      # get country info given some params
      def country(options = {:name => "", :num =>"", :alpha2s => "", :alpha3s => ""})
        result = []
        COUNTRY_NUMS.zip(translate_countries, COUNTRY_ALPHA2S, COUNTRY_ALPHA3S).each do |country|
          if find_country(options, country)
           result = country
          end
        end
        result
      end

      #get country name
      def country_name(options = {:name => "", :num =>"", :alpha2s => "", :alpha3s => ""})
        country(options)[1]
      end

      #get country cod
      def country_cod(options = {:name => "", :num =>"", :alpha2s => "", :alpha3s => ""})
        country(options)[0]
      end

      #get country alpha2s
      def country_alpha2s(options = {:name => "", :num =>"", :alpha2s => "", :alpha3s => ""})
        country(options)[2]
      end

      #get country alpha3s
      def country_alpha3s(options = {:name => "", :num =>"", :alpha2s => "", :alpha3s => ""})
        country(options)[3]
      end

      private
      def translate_countries
        COUNTRY_ALPHA2S.zip(COUNTRY_NAMES).map do |code, name|
          I18n.t(code, :scope => :countries, :default => name)
        end
      end

      def find_country(options, country)
        options[:num].to_i.eql?(country[0]) || options[:name].eql?(country[1]) || options[:alpha2s].eql?(country[2]) || options[:alpha3s].eql?(country[3])
      end
    end
  end
#end



lib/country_definitions.rb
# encoding: utf-8

COUNTRY_NAMES   = [
    'Afghanistan',
    'Aland Islands',
    'Albania',
    'Algeria',
    'American Samoa',
    'Andorra',
    'Angola',
    'Anguilla',
    'Antarctica',
    'Antigua and Barbuda',
    'Argentina',
    'Armenia',
    'Aruba',
    'Australia',
    'Austria',
    'Azerbaijan',
    'Bahamas',
    'Bahrain',
    'Bangladesh',
    'Barbados',
    'Belarus',
    'Belgium',
    'Belize',
    'Benin',
    'Bermuda',
    'Bhutan',
    'Bolivia, Plurinational State of',
    'Bonaire, Sint Eustatius and Saba',
    'Bosnia and Herzegovina',
    'Botswana',
    'Bouvet Island',
    'Brazil',
    'British Indian Ocean Territory',
    'Brunei Darussalam',
    'Bulgaria',
    'Burkina Faso',
    'Burundi',
    'Cambodia',
    'Cameroon',
    'Canada',
    'Cape Verde',
    'Cayman Islands',
    'Central African Republic',
    'Chad',
    'Chile',
    'China',
    'Christmas Island',
    'Cocos (Keeling) Islands',
    'Colombia',
    'Comoros',
    'Congo',
    'Congo, the Democratic Republic of the',
    'Cook Islands',
    'Costa Rica',
    'CÃ´te d\'Ivoire',
    'Croatia',
    'Cuba',
    'CuraÃ§ao',
    'Cyprus',
    'Czech Republic',
    'Denmark',
    'Djibouti',
    'Dominica',
    'Dominican Republic',
    'Ecuador',
    'Egypt',
    'El Salvador',
    'Equatorial Guinea',
    'Eritrea',
    'Estonia',
    'Ethiopia',
    'Falkland Islands (Malvinas)',
    'Faroe Islands',
    'Fiji',
    'Finland',
    'France',
    'French Guiana',
    'French Polynesia',
    'French Southern Territories',
    'Gabon',
    'Gambia',
    'Georgia',
    'Germany',
    'Ghana',
    'Gibraltar',
    'Greece',
    'Greenland',
    'Grenada',
    'Guadeloupe',
    'Guam',
    'Guatemala',
    'Guernsey',
    'Guinea',
    'Guinea-Bissau',
    'Guyana',
    'Haiti',
    'Heard Island and McDonald Islands',
    'Holy See (Vatican City State)',
    'Honduras',
    'Hong Kong',
    'Hungary',
    'Iceland',
    'India',
    'Indonesia',
    'Iran, Islamic Republic of',
    'Iraq',
    'Ireland',
    'Isle of Man',
    'Israel',
    'Italy',
    'Jamaica',
    'Japan',
    'Jersey',
    'Jordan',
    'Kazakhstan',
    'Kenya',
    'Kiribati',
    'Korea, Democratic People\'s Republic of',
    'Korea, Republic of',
    'Kuwait',
    'Kyrgyzstan',
    'Lao People\'s Democratic Republic',
    'Latvia',
    'Lebanon',
    'Lesotho',
    'Liberia',
    'Libyan Arab Jamahiriya',
    'Liechtenstein',
    'Lithuania',
    'Luxembourg',
    'Macao',
    'Macedonia, the former Yugoslav Republic of',
    'Madagascar',
    'Malawi',
    'Malaysia',
    'Maldives',
    'Mali',
    'Malta',
    'Marshall Islands',
    'Martinique',
    'Mauritania',
    'Mauritius',
    'Mayotte',
    'Mexico',
    'Micronesia, Federated States of',
    'Moldova, Republic of',
    'Monaco',
    'Mongolia',
    'Montenegro',
    'Montserrat',
    'Morocco',
    'Mozambique',
    'Myanmar',
    'Namibia',
    'Nauru',
    'Nepal',
    'Netherlands',
    'New Caledonia',
    'New Zealand',
    'Nicaragua',
    'Niger',
    'Nigeria',
    'Niue',
    'Norfolk Island',
    'Northern Mariana Islands',
    'Norway',
    'Oman',
    'Pakistan',
    'Palau',
    'Palestinian Territory, Occupied',
    'Panama',
    'Papua New Guinea',
    'Paraguay',
    'Peru',
    'Philippines',
    'Pitcairn',
    'Poland',
    'Portugal',
    'Puerto Rico',
    'Qatar',
    'Reunion',
    'Romania',
    'Russian Federation',
    'Rwanda',
    'Saint BarthÃ©lemy',
    'Saint Helena, Ascension and Tristan da Cunha',
    'Saint Kitts and Nevis',
    'Saint Lucia',
    'Saint Martin (French part)',
    'Saint Pierre and Miquelon',
    'Saint Vincent and the Grenadines',
    'Samoa',
    'San Marino',
    'Sao Tome and Principe',
    'Saudi Arabia',
    'Senegal',
    'Serbia',
    'Seychelles',
    'Sierra Leone',
    'Singapore',
    'Sint Maarten (Dutch part)',
    'Slovakia',
    'Slovenia',
    'Solomon Islands',
    'Somalia',
    'South Africa',
    'South Georgia and the South Sandwich Islands',
    'South Sudan',
    'Spain',
    'Sri Lanka',
    'Sudan',
    'Suriname',
    'Svalbard and Jan Mayen',
    'Swaziland',
    'Sweden',
    'Switzerland',
    'Syrian Arab Republic',
    'Taiwan, Province of China',
    'Tajikistan',
    'Tanzania, United Republic of',
    'Thailand',
    'Timor-Leste',
    'Togo',
    'Tokelau',
    'Tonga',
    'Trinidad and Tobago',
    'Tunisia',
    'Turkey',
    'Turkmenistan',
    'Turks and Caicos Islands',
    'Tuvalu',
    'Uganda',
    'Ukraine',
    'United Arab Emirates',
    'United Kingdom',
    'United States',
    'United States Minor Outlying Islands',
    'Uruguay',
    'Uzbekistan',
    'Vanuatu',
    'Venezuela, Bolivarian Republic of',
    'Viet Nam',
    'Virgin Islands, British',
    'Virgin Islands, U.S.',
    'Wallis and Futuna',
    'Western Sahara',
    'Yemen',
    'Zambia',
    'Zimbabwe',
]

COUNTRY_ALPHA2S = [
    'AF',
    'AX',
    'AL',
    'DZ',
    'AS',
    'AD',
    'AO',
    'AI',
    'AQ',
    'AG',
    'AR',
    'AM',
    'AW',
    'AU',
    'AT',
    'AZ',
    'BS',
    'BH',
    'BD',
    'BB',
    'BY',
    'BE',
    'BZ',
    'BJ',
    'BM',
    'BT',
    'BO',
    'BQ',
    'BA',
    'BW',
    'BV',
    'BR',
    'IO',
    'BN',
    'BG',
    'BF',
    'BI',
    'KH',
    'CM',
    'CA',
    'CV',
    'KY',
    'CF',
    'TD',
    'CL',
    'CN',
    'CX',
    'CC',
    'CO',
    'KM',
    'CG',
    'CD',
    'CK',
    'CR',
    'CI',
    'HR',
    'CU',
    'CW',
    'CY',
    'CZ',
    'DK',
    'DJ',
    'DM',
    'DO',
    'EC',
    'EG',
    'SV',
    'GQ',
    'ER',
    'EE',
    'ET',
    'FK',
    'FO',
    'FJ',
    'FI',
    'FR',
    'GF',
    'PF',
    'TF',
    'GA',
    'GM',
    'GE',
    'DE',
    'GH',
    'GI',
    'GR',
    'GL',
    'GD',
    'GP',
    'GU',
    'GT',
    'GG',
    'GN',
    'GW',
    'GY',
    'HT',
    'HM',
    'VA',
    'HN',
    'HK',
    'HU',
    'IS',
    'IN',
    'ID',
    'IR',
    'IQ',
    'IE',
    'IM',
    'IL',
    'IT',
    'JM',
    'JP',
    'JE',
    'JO',
    'KZ',
    'KE',
    'KI',
    'KP',
    'KR',
    'KW',
    'KG',
    'LA',
    'LV',
    'LB',
    'LS',
    'LR',
    'LY',
    'LI',
    'LT',
    'LU',
    'MO',
    'MK',
    'MG',
    'MW',
    'MY',
    'MV',
    'ML',
    'MT',
    'MH',
    'MQ',
    'MR',
    'MU',
    'YT',
    'MX',
    'FM',
    'MD',
    'MC',
    'MN',
    'ME',
    'MS',
    'MA',
    'MZ',
    'MM',
    'NA',
    'NR',
    'NP',
    'NL',
    'NC',
    'NZ',
    'NI',
    'NE',
    'NG',
    'NU',
    'NF',
    'MP',
    'NO',
    'OM',
    'PK',
    'PW',
    'PS',
    'PA',
    'PG',
    'PY',
    'PE',
    'PH',
    'PN',
    'PL',
    'PT',
    'PR',
    'QA',
    'RE',
    'RO',
    'RU',
    'RW',
    'BL',
    'SH',
    'KN',
    'LC',
    'MF',
    'PM',
    'VC',
    'WS',
    'SM',
    'ST',
    'SA',
    'SN',
    'RS',
    'SC',
    'SL',
    'SG',
    'SX',
    'SK',
    'SI',
    'SB',
    'SO',
    'ZA',
    'GS',
    'SS',
    'ES',
    'LK',
    'SD',
    'SR',
    'SJ',
    'SZ',
    'SE',
    'CH',
    'SY',
    'TW',
    'TJ',
    'TZ',
    'TH',
    'TL',
    'TG',
    'TK',
    'TO',
    'TT',
    'TN',
    'TR',
    'TM',
    'TC',
    'TV',
    'UG',
    'UA',
    'AE',
    'GB',
    'US',
    'UM',
    'UY',
    'UZ',
    'VU',
    'VE',
    'VN',
    'VG',
    'VI',
    'WF',
    'EH',
    'YE',
    'ZM',
    'ZW',
]

COUNTRY_ALPHA3S = [
    'AFG',
    'ALA',
    'ALB',
    'DZA',
    'ASM',
    'AND',
    'AGO',
    'AIA',
    'ATA',
    'ATG',
    'ARG',
    'ARM',
    'ABW',
    'AUS',
    'AUT',
    'AZE',
    'BHS',
    'BHR',
    'BGD',
    'BRB',
    'BLR',
    'BEL',
    'BLZ',
    'BEN',
    'BMU',
    'BTN',
    'BOL',
    'BES',
    'BIH',
    'BWA',
    'BVT',
    'BRA',
    'IOT',
    'BRN',
    'BGR',
    'BFA',
    'BDI',
    'KHM',
    'CMR',
    'CAN',
    'CPV',
    'CYM',
    'CAF',
    'TCD',
    'CHL',
    'CHN',
    'CXR',
    'CCK',
    'COL',
    'COM',
    'COG',
    'COD',
    'COK',
    'CRI',
    'CIV',
    'HRV',
    'CUB',
    'CUW',
    'CYP',
    'CZE',
    'DNK',
    'DJI',
    'DMA',
    'DOM',
    'ECU',
    'EGY',
    'SLV',
    'GNQ',
    'ERI',
    'EST',
    'ETH',
    'FLK',
    'FRO',
    'FJI',
    'FIN',
    'FRA',
    'GUF',
    'PYF',
    'ATF',
    'GAB',
    'GMB',
    'GEO',
    'DEU',
    'GHA',
    'GIB',
    'GRC',
    'GRL',
    'GRD',
    'GLP',
    'GUM',
    'GTM',
    'GGY',
    'GIN',
    'GNB',
    'GUY',
    'HTI',
    'HMD',
    'VAT',
    'HND',
    'HKG',
    'HUN',
    'ISL',
    'IND',
    'IDN',
    'IRN',
    'IRQ',
    'IRL',
    'IMN',
    'ISR',
    'ITA',
    'JAM',
    'JPN',
    'JEY',
    'JOR',
    'KAZ',
    'KEN',
    'KIR',
    'PRK',
    'KOR',
    'KWT',
    'KGZ',
    'LAO',
    'LVA',
    'LBN',
    'LSO',
    'LBR',
    'LBY',
    'LIE',
    'LTU',
    'LUX',
    'MAC',
    'MKD',
    'MDG',
    'MWI',
    'MYS',
    'MDV',
    'MLI',
    'MLT',
    'MHL',
    'MTQ',
    'MRT',
    'MUS',
    'MYT',
    'MEX',
    'FSM',
    'MDA',
    'MCO',
    'MNG',
    'MNE',
    'MSR',
    'MAR',
    'MOZ',
    'MMR',
    'NAM',
    'NRU',
    'NPL',
    'NLD',
    'NCL',
    'NZL',
    'NIC',
    'NER',
    'NGA',
    'NIU',
    'NFK',
    'MNP',
    'NOR',
    'OMN',
    'PAK',
    'PLW',
    'PSE',
    'PAN',
    'PNG',
    'PRY',
    'PER',
    'PHL',
    'PCN',
    'POL',
    'PRT',
    'PRI',
    'QAT',
    'REU',
    'ROU',
    'RUS',
    'RWA',
    'BLM',
    'SHN',
    'KNA',
    'LCA',
    'MAF',
    'SPM',
    'VCT',
    'WSM',
    'SMR',
    'STP',
    'SAU',
    'SEN',
    'SRB',
    'SYC',
    'SLE',
    'SGP',
    'SXM',
    'SVK',
    'SVN',
    'SLB',
    'SOM',
    'ZAF',
    'SGS',
    'SSD',
    'ESP',
    'LKA',
    'SDN',
    'SUR',
    'SJM',
    'SWZ',
    'SWE',
    'CHE',
    'SYR',
    'TWN',
    'TJK',
    'TZA',
    'THA',
    'TLS',
    'TGO',
    'TKL',
    'TON',
    'TTO',
    'TUN',
    'TUR',
    'TKM',
    'TCA',
    'TUV',
    'UGA',
    'UKR',
    'ARE',
    'GBR',
    'USA',
    'UMI',
    'URY',
    'UZB',
    'VUT',
    'VEN',
    'VNM',
    'VGB',
    'VIR',
    'WLF',
    'ESH',
    'YEM',
    'ZMB',
    'ZWE',
]

COUNTRY_NUMS    = [
    4,
    248,
    8,
    12,
    16,
    20,
    24,
    660,
    10,
    28,
    32,
    51,
    533,
    36,
    40,
    31,
    44,
    48,
    50,
    52,
    112,
    56,
    84,
    204,
    60,
    64,
    68,
    535,
    70,
    72,
    74,
    76,
    86,
    96,
    100,
    854,
    108,
    116,
    120,
    124,
    132,
    136,
    140,
    148,
    152,
    156,
    162,
    166,
    170,
    174,
    178,
    180,
    184,
    188,
    384,
    191,
    192,
    531,
    196,
    203,
    208,
    262,
    212,
    214,
    218,
    818,
    222,
    226,
    232,
    233,
    231,
    238,
    234,
    242,
    246,
    250,
    254,
    258,
    260,
    266,
    270,
    268,
    276,
    288,
    292,
    300,
    304,
    308,
    312,
    316,
    320,
    831,
    324,
    624,
    328,
    332,
    334,
    336,
    340,
    344,
    348,
    352,
    356,
    360,
    364,
    368,
    372,
    833,
    376,
    380,
    388,
    392,
    832,
    400,
    398,
    404,
    296,
    408,
    410,
    414,
    417,
    418,
    428,
    422,
    426,
    430,
    434,
    438,
    440,
    442,
    446,
    807,
    450,
    454,
    458,
    462,
    466,
    470,
    584,
    474,
    478,
    480,
    175,
    484,
    583,
    498,
    492,
    496,
    499,
    500,
    504,
    508,
    104,
    516,
    520,
    524,
    528,
    540,
    554,
    558,
    562,
    566,
    570,
    574,
    580,
    578,
    512,
    586,
    585,
    275,
    591,
    598,
    600,
    604,
    608,
    612,
    616,
    620,
    630,
    634,
    638,
    642,
    643,
    646,
    652,
    654,
    659,
    662,
    663,
    666,
    670,
    882,
    674,
    678,
    682,
    686,
    688,
    690,
    694,
    702,
    534,
    703,
    705,
    90,
    706,
    710,
    239,
    728,
    724,
    144,
    729,
    740,
    744,
    748,
    752,
    756,
    760,
    158,
    762,
    834,
    764,
    626,
    768,
    772,
    776,
    780,
    788,
    792,
    795,
    796,
    798,
    800,
    804,
    784,
    826,
    840,
    581,
    858,
    860,
    548,
    862,
    704,
    92,
    850,
    876,
    732,
    887,
    894,
    716,
]

